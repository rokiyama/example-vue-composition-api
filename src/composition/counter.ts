import { ref } from "@vue/composition-api";

export function useCount() {
  const counter = ref(0);
  const increment = () => {
    counter.value++;
  };
  const decrement = () => {
    counter.value--;
  };
  return {
    counter,
    increment,
    decrement
  };
}
