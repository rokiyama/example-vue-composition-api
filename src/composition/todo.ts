import { reactive, computed, toRefs } from "@vue/composition-api";

export function useTodo() {
  const state = reactive<{ todos: string[] }>({
    todos: []
  });
  const length = computed(() => state.todos.length);
  const addTodo = (item: string) => {
    state.todos.push(item);
  };
  const deleteTodo = (index: number) => {
    state.todos.splice(index, 1);
  };
  return {
    ...toRefs(state),
    length,
    addTodo,
    deleteTodo
  };
}
