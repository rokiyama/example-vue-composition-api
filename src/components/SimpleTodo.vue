<template>
  <div class="count">
    <p>
      <input id="todo-input" v-model="text" />
      <button class="add-btn" @click="onSubmit">Add</button>
    </p>
    <p>{{ length }} todo{{ length > 1 ? "s" : "" }}</p>
    <ul>
      <li v-for="(task, i) in todos" :key="i">
        <p>
          {{ task }}
          <button class="delete-btn" @click="deleteTodo(i)">Delete</button>
        </p>
      </li>
    </ul>
  </div>
</template>

<script lang="ts">
import { createComponent, ref } from "@vue/composition-api";
import { useTodo } from "@/composition/todo";

export default createComponent({
  setup() {
    const text = ref("");
    const { todos, length, addTodo, deleteTodo } = useTodo();
    const onSubmit = () => {
      addTodo(text.value);
      text.value = "";
    };
    return {
      text,
      todos,
      length,
      addTodo,
      deleteTodo,
      onSubmit
    };
  }
});
</script>
