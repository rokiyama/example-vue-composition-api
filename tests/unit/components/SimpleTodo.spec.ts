import VueCompositionApi, { ref } from "@vue/composition-api";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import SimpleTodo from "@/components/SimpleTodo.vue";
import * as composition from "@/composition/todo";

const localVue = createLocalVue();
localVue.use(VueCompositionApi);

let addTodoMock: jest.Mock;
let deleteTodoMock: jest.Mock;

describe("SimpleTodo.vue", () => {
  beforeEach(() => {
    jest.mock("@/composition/todo");
    addTodoMock = jest.fn();
    deleteTodoMock = jest.fn();
    const TODOS = ["todo1", "todo2", "todo3"];
    jest.spyOn(composition, "useTodo").mockReturnValue({
      todos: ref(TODOS),
      length: ref(TODOS.length),
      addTodo: addTodoMock,
      deleteTodo: deleteTodoMock
    });
  });
  it("renders initial html", () => {
    const wrapper = shallowMount(SimpleTodo, {
      localVue
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("correctly call addTodo when Add button is clicked", () => {
    const wrapper = shallowMount(SimpleTodo, {
      localVue
    });
    wrapper.find("#todo-input").setValue("todo4");
    wrapper.find(".add-btn").trigger("click");
    expect(addTodoMock).toHaveBeenCalledWith("todo4");
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("correctly call deleteTodo when Delete button is clicked", () => {
    const wrapper = shallowMount(SimpleTodo, {
      localVue
    });
    const INDEX = 1;
    wrapper
      .findAll(".delete-btn")
      .at(INDEX)
      .trigger("click");
    expect(deleteTodoMock).toHaveBeenCalledWith(INDEX);
  });
});
