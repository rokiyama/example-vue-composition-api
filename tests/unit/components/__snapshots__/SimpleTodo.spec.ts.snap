// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTodo.vue correctly call addTodo when Add button is clicked 1`] = `
<div class="count">
  <p><input id="todo-input"> <button class="add-btn">Add</button></p>
  <p>3 todos</p>
  <ul>
    <li>
      <p>
        todo1
        <button class="delete-btn">Delete</button></p>
    </li>
    <li>
      <p>
        todo2
        <button class="delete-btn">Delete</button></p>
    </li>
    <li>
      <p>
        todo3
        <button class="delete-btn">Delete</button></p>
    </li>
  </ul>
</div>
`;

exports[`SimpleTodo.vue renders initial html 1`] = `
<div class="count">
  <p><input id="todo-input"> <button class="add-btn">Add</button></p>
  <p>3 todos</p>
  <ul>
    <li>
      <p>
        todo1
        <button class="delete-btn">Delete</button></p>
    </li>
    <li>
      <p>
        todo2
        <button class="delete-btn">Delete</button></p>
    </li>
    <li>
      <p>
        todo3
        <button class="delete-btn">Delete</button></p>
    </li>
  </ul>
</div>
`;
